### PR TITLE
Intel 8080 parallel interface HIL, Bus library with I2C, SPI and Intel 8080 parallel interface implementation

### PR DESCRIPTION
**Pull request by alexandruradovici**
### Pull Request Overview

Several hardware components work with the same protocol over several buses. For instance, sensors like l3gd20, lsm303 work over I2C and SPI, screens like ST7735 and ST7789H2 work over SPI or Intel 8080 parallel interface. Using the Bus library, capsules for these devices will work regardless of which hardware communication bus is used.

This PR
  * adds screen support for stm32f412gdiscovery
  * adds FSMC support for stm32f4xx
  * a Intel 8080 parallel interface (Bus8080) HIL
  * changes st7735 driver to st77xx single driver for ST7735 , ST7789H2 and LS016B8UY (reversed engineered, no documentation available) over the new Bus HIL
  * fixes the screen HIL and capsule (adds a `write_continue` function to the screen trait)

#### Intel 8080 parallel interface HIL

```rust
pub trait Bus8080<'a> {
    /// Set the address to write to
    fn set_addr(&self, addr_width: BusWidth, addr: usize) -> ReturnCode;

    /// Write data items to the previously set address
    fn write(&self, data_width: BusWidth, buffer: &'static mut [u8], len: usize) -> ReturnCode;

    /// Read data items from the previously set address
    fn read(&self, data_width: BusWidth, buffer: &'static mut [u8], len: usize) -> ReturnCode;

    fn set_client(&self, client: &'a dyn Client);
}

pub trait Client {
    /// Called when set_addr, write or read are complete
    ///
    /// set_address does not return a buffer
    /// write and read return a buffer
    /// len should be set to the number of data elements written
    fn command_complete(&self, buffer: Option<&'static mut [u8]>, len: usize);
}
```

The Bus8080 HIL is required as FSMC seems to be specific to STMs and is used to connect devices that support Parallel 8080 interface. Several screens work with this interface.

#### Bus Librarry

```rust
pub trait Bus<'a> {
    /// Set the address to write to
    ///
    /// If the underlaying bus does not support addresses (eg UART)
    /// this function returns ENOSUPPORT
    fn set_addr(&self, addr_width: BusWidth, addr: usize) -> ReturnCode;

    /// Write data items to the previously set address
    ///
    /// data_width specifies the encoding of the data items placed in the buffer
    /// len specifies the number of data items (the number of bytes is len * data_width.width_in_bytes)
    fn write(&self, data_width: BusWidth, buffer: &'static mut [u8], len: usize) -> ReturnCode;

    /// Read data items from the previously set address
    ///
    /// data_width specifies the encoding of the data items placed in the buffer
    /// len specifies the number of data items (the number of bytes is len * data_width.width_in_bytes)
    fn read(&self, data_width: BusWidth, buffer: &'static mut [u8], len: usize) -> ReturnCode;

    fn set_client(&self, client: &'static dyn Client);
}

pub trait Client {
    /// Called when set_addr, write or read are complete
    ///
    /// set_address does not return a buffer
    /// write and read return a buffer
    /// len should be set to the number of data elements written
    fn command_complete(&self, buffer: Option<&'static mut [u8]>, len: usize);
}
```

#### Example

Example for ST7735 screen before bus (spi dependent driver)
```rust
//! ```rust
 let tft = components::st7755::ST7735Component::new(alarm_mux).finalize(
     components::st7735_component_helper!(
         // spi type
         stm32f4xx::spi::Spi,
         // chip select
         stm32f4xx::gpio::PinId::PE03,
         // spi mux
         spi_mux,
         // timer type
         stm32f4xx::tim2::Tim2,
         // dc pin
         stm32f4xx::gpio::PinId::PA00.get_pin().as_ref().unwrap(),
         // reset pin
         stm32f4xx::gpio::PinId::PA00.get_pin().as_ref().unwrap()
     )
 );
```

Example for ST7735 screen after bus (bus dependent library)
```rust
let bus = components::bus::Bus8080BusComponent::new().finalize(
    components::bus8080_bus_component_helper!(
        // bus type
        stm32f412g::fsmc::Fsmc,
        // bus
        &stm32f412g::fsmc::FSMC
    ),
);

let tft = components::st77xx::ST77XXComponent::new(mux_alarm).finalize(
    components::st77xx_component_helper!(
        // screen
        &capsules::st77xx::ST7789H2,
        // bus type
        capsules::bus::Bus8080Bus<'static, stm32f412g::fsmc::Fsmc>,
        // bus
        &bus,
        // timer type
        stm32f412g::tim2::Tim2,
        // pin type
        stm32f412g::gpio::Pin,
        // dc pin (optional)
        None,
        // reset pin
        stm32f412g::gpio::PinId::PD11.get_pin().as_ref().unwrap()
    ),
);
```

### Testing Strategy

This pull request was tested using ST7735 and ST7789H2 screens using nano33ble and stm32f412gdiscovery

### TODO or Help Wanted

This pull request still needs some help for the FSMC. I have no idea how to make the write asynchronous, there is a FSMC interrupt but I'm not sure how to call it. One option would be DMA, but it does not seem to be available for FSMC. Any idea would be very useful, as the screen for stm32f412gdiscovery is connected the FSMC.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Comment by alexandruradovici** [comment]
I updated the bus and the st77xx drivers to use generics, it seems to reduce the code size by about 600 B. The code does look a little more complex.

**Comment by bradjc** [comment]
I'm a little wary of adding another layer of abstraction (the Bus interface). I see what this would provide, but:
- I'm concerned this won't actually work, due to potential corner cases in different implementations between different chips (both MCUs and sensors). For example, implementing `Bus` for one chip might work for one sensor, but not for a different sensor. To capture these potential subtleties the Bus trait would have to get a lot more complicated.
- This will be very hard to test, since a chip capsule developer is very unlikely to have hardware to test the chip with both interfaces. This will almost ensure that we have capsules that look like they provide more support than has ever been tested.
- This adds another layer of abstraction which adds more complexity when debugging or understanding the code. And, we haven't run into the issue where a board uses an existing capsule but with a different bus.

**Comment by alexandruradovici** [comment]
I think this discussion is going to be longer, so let me split this into several PR. I'll leave the bus here, add the FMSC and screens to other PRs.

**Comment by alexandruradovici** [comment]
> I'm a little wary of adding another layer of abstraction (the Bus interface). I see what this would provide, but:
> 
>     * I'm concerned this won't actually work, due to potential corner cases in different implementations between different chips (both MCUs and sensors). For example, implementing `Bus` for one chip might work for one sensor, but not for a different sensor. To capture these potential subtleties the Bus trait would have to get a lot more complicated.

If the chips use some specific functions (or timing) for a specific protocol, I think they should the protocol capsules directly.

> 
>     * This will be very hard to test, since a chip capsule developer is very unlikely to have hardware to test the chip with both interfaces. This will almost ensure that we have capsules that look like they provide more support than has ever been tested.
> 

I don;t thnik this is a problem, components for chips might be defined for several buses. The developer would simply not write the component that he did not test.

```rust
#[macro_export]
macro_rules! st7735_i2c_component_helper {
  // ...
}

macro_rules! st7735_spi_component_helper {
  // ...
}

macro_rules! st7735_intel8080_component_helper {
  // ...
}
```

>     * This adds another layer of abstraction which adds more complexity when debugging or understanding the code. And, we haven't run into the issue where a board uses an existing capsule but with a different bus.

I have run into it when writing drivers for st77xxx screens. The LSM sensors have the same issue, they support both I2C and SPI.

File: boards/stm32f412gdiscovery/src/main.rs
Comment (alistair23) [inline boards/stm32f412gdiscovery/src/main.rs:10]: Why is this needed?

**Comment by alistair23** [comment]
I agree with Brad here, at first I thought this would be really useful, but I think there will just be too many corner cases for this to be reliable.

For devices where this is the case couldn't you just implement both I2C and SPI for the device and then call generic functions? That will result in supporting both protocols with little work.

File: boards/stm32f412gdiscovery/src/main.rs
Comment (alexandruradovici) [reply boards/stm32f412gdiscovery/src/main.rs:10]: I left it there by mistake, deleting it.

**Comment by alexandruradovici** [comment]
> I agree with Brad here, at first I thought this would be really useful, but I think there will just be too many corner cases for this to be reliable.
> 
I understand your point.
> For devices where this is the case couldn't you just implement both I2C and SPI for the device and then call generic functions? That will result in supporting both protocols with little work.

What about adding this as a library in libraries, I think I can use it for several capsules.

**Comment by alistair23** [comment]
I don't think it needs to be a library, it won't be very long.

**Comment by alexandruradovici** [comment]
It's not long, I just need to use it in several capsules and I am not sure where to place the code.

**Comment by alistair23** [comment]
What I'm picturing (which could be wrong) isn't too much code duplication that it's probably not an issue. Up to you though

**Comment by bradjc** [comment]
The core team discussed how to move forward with this and we think the best plan is to make `Bus` a library in capsules/. Ideally it would have a more specific name for the protocol on top of the physical bus (if there is a reasonable name). This main difference between this and a HIL is just that the implementations for different physical busses would be in capsules/ rather than chips/.

We also think that separating the transport layer from the operational logic on a chip-by-chip basis would work well, but if there are multiple chips with the same interface protocol then consolidating the interfaces into a library makes a lot of sense.

**Comment by alexandruradovici** [comment]
> The core team discussed how to move forward with this and we think the best plan is to make `Bus` a library in capsules/. Ideally it would have a more specific name for the protocol on top of the physical bus (if there is a reasonable name). This main difference between this and a HIL is just that the implementations for different physical busses would be in capsules/ rather than chips/.
> 

Ok, that makes sense. I'm not sure how to proceed. Should there be a trait declared in capsules?

> We also think that separating the transport layer from the operational logic on a chip-by-chip basis would work well, but if there are multiple chips with the same interface protocol then consolidating the interfaces into a library makes a lot of sense.

**Comment by alexandruradovici** [comment]
> > The core team discussed how to move forward with this and we think the best plan is to make `Bus` a library in capsules/. Ideally it would have a more specific name for the protocol on top of the physical bus (if there is a reasonable name). This main difference between this and a HIL is just that the implementations for different physical busses would be in capsules/ rather than chips/.
> 
> Ok, that makes sense. I'm not sure how to proceed. Should there be a trait declared in capsules?
> 
> > We also think that separating the transport layer from the operational logic on a chip-by-chip basis would work well, but if there are multiple chips with the same interface protocol then consolidating the interfaces into a library makes a lot of sense.

I'd love to close this as soon as possible, but @bradjc I would really need some clarifications on what you have in mind. Thank you.

**Comment by alexandruradovici** [comment]
Hi guys, just a kind reminder for some feedback here about the bus library? Thanks.

**Comment by hudson-ayers** [comment]
@alexandruradovici Here is a link to the discussion in the working group meeting a few weeks back: https://github.com/tock/tock/blob/master/doc/wg/core/notes/core-notes-2020-08-07.md#generic-bus-hil . I think the desired changes include declaring the trait in one file in `capsules/` (rather than in `kernel/hil') and then implementing the trait in other files in `capsules/`. Is there a reason that is problematic for your use case?

**Comment by alexandruradovici** [comment]
I made the change from the HIL to the library Trait and updated the PR description.

**Comment by alexandruradovici** [comment]
Hi, any feedback here? It would really help us to close this. Thanks.

File: chips/stm32f4xx/src/fsmc.rs
Comment (hudson-ayers) [inline chips/stm32f4xx/src/fsmc.rs:149]: I don't like this practice of treating an arbitrary memory location as though it is an `FsmcBank`. This could easily lead to memory unsafety if these slots were used. If empty slots are needed in the FSMC array, the array should hold `Option<StaticRef<FsmcBank>>`

File: kernel/src/hil/screen.rs
Comment (hudson-ayers) [inline kernel/src/hil/screen.rs:196]: I am not clear on why this function is being added to the trait. What can lead to a write beginning, but not being finished? How should clients know when `write_continue()` needs to be called?

**Review by hudson-ayers — CHANGES_REQUESTED** [review]
A couple comments. I think this is close to being merged, though I would like another set of eyes on the changes in `kernel/src/hil/`.

File: chips/stm32f4xx/src/fsmc.rs
Comment (alexandruradovici) [reply chips/stm32f4xx/src/fsmc.rs:149]: Thanks for the tip, I modified it.

File: kernel/src/hil/screen.rs
Comment (alexandruradovici) [reply kernel/src/hil/screen.rs:196]: The frame buffer data might be larger than the allocated RAM, for instance when performing a fill. 

- `write` sets the memory address at the beginning of the screen frame window
- `write_continue` does the same thing as `write`, but does not reset the frame windows position

File: chips/stm32f4xx/src/fsmc.rs
Comment (hudson-ayers) [inline chips/stm32f4xx/src/fsmc.rs]: Should these be labeled BANK0..BANK3 instead of BANK1..BANK4 to match the indicies in the array in which they are contained? It seems odd that programming the bank labeled BANK1 requires passing a `bank_id` of `0` to `write_data()`

File: kernel/src/hil/screen.rs
Comment (hudson-ayers) [reply kernel/src/hil/screen.rs:196]: That makes sense. Can you add that description to the comments in this file? I initially misread this as `write_continue()` somehow resuming a paused write.

Comment (alexandruradovici) [reply kernel/src/hil/screen.rs]: Done

File: chips/stm32f4xx/src/fsmc.rs
Comment (alexandruradovici) [reply chips/stm32f4xx/src/fsmc.rs]: I added an enum for the Banks instead of using an index.

**Review by hudson-ayers — APPROVED** [review]
Looks good to me, would be great to have one more set of eyes on the HIL changes

**Comment by alevy** [comment]
This looks good to me. I suggest we wait until after 1.6 to merge (#2030), and then merge immediately.

**Review by alevy — DISMISSED** [review]

**Review by hudson-ayers — APPROVED** [review]
bors r+

**Pull request merged**
